Addition to Filesystem Benchmark Testing

uma-sethuraman opened the pull request:
Hello @wjhun @francescolavra, this pull request is for issue #1166. As mentioned in the comments of that issue, we added a filesystem stress test to `test/runtime/write.c` which creates a large number of files, writes data to each of those files, syncs the filesystem, deletes all the files, and then syncs again. Please let us know what you think of the test and if there's anything we should change or add.

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +618,59 @@ static void usage(const char *program_name)
+    }
+
+    /* sync the filesystem */
+    syscall(SYS_sync); 
```
Couldn't you just call the sync() function here?

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +618,59 @@ static void usage(const char *program_name)
+        close(fds[i]);
+
+    /* sync the filesystem again */
+    syscall(SYS_sync); 
```
Same here, you could just call the sync() function

francescolavra commented on test/runtime/write.c:
```diff
@@ -12,6 +12,7 @@
 #include <sys/time.h>
 #include <sys/eventfd.h>
 #include <sys/stat.h>
+#include <unix/syscalls.h>
```
This include shouldn't be needed anymore once `syscall(SYS_sync)` is replaced by `sync()`

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +618,59 @@ static void usage(const char *program_name)
+        fds[i] = fd;
+        if (fd < 0) {
+            printf("open error\n");
+            perror("open");
```
The printf above is redundant, since there is the perror() call

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +618,59 @@ static void usage(const char *program_name)
+
+    /* Deleting all files we just created */
+    for (i = 0; i < num_files; i++) 
+        close(fds[i]);
```
Closing a file does not delete it

samantha3pen replied:
@francescolavra we used remove() and added a check to ensure the files were properly removed. Is this synonymous to deleting a file?

francescolavra replied:
Yes, remove() can be used to delete a file, but if you check the return value of remove() in your code you will see that it's an error value for most (all?) files, so it's not being used correctly and is not actually deleting those files.

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +618,59 @@ static void usage(const char *program_name)
+        _READ(buf, BUFLEN);
+
+        if (rv == 0)
+            writetest_debug("empty source file\n");
```
This writetest_debug() call is not really useful, since all files will be empty and this would just pollute the debug logs with 1000 identical lines

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +618,59 @@ static void usage(const char *program_name)
+        if (rv >= BUFLEN)
+            rv = BUFLEN - 1;
+        buf[rv] = '\0';
+        writetest_debug("original: \"%s\"\n", buf);
```
Same here

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +617,67 @@ static void usage(const char *program_name)
+        sprintf(name, "fs_stress_test_%d", i);
+        int fd = open(name, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR);
+        fds[i] = fd;
+        fd_names[i] = name;
```
`name` is a variable that lives only within the `for()` loop, you cannot assign its address to fd_names[i] and then use fd_names[i] outside the `for()` loop.

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +617,67 @@ static void usage(const char *program_name)
+
+        if (rv >= BUFLEN)
+            rv = BUFLEN - 1;
+        buf[rv] = '\0';
```
What's the purpose of the 3 lines above?

uma-sethuraman replied:
We actually were using these lines with our `writetest_debug` statements, but since we got rid of those, these lines are no longer needed. We deleted these three lines.

francescolavra reviewed (commented):
It would be good for the FS stress test to print on the standard output how long it takes to execute, and more specifically how long it takes up to the first `sync()` call, and then how long the remaining part take, so that we have some statistics shown on the terminal when the test runs, similarly to how it's done in bulk_write_test().

uma-sethuraman commented:
@francescolavra we updated the test to print out statistics similar to bulk_write_test() and have made the changes you mentioned earlier. We also fixed the errors we were getting from the remove() statement. Let us know if there's anything else we should change.

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +617,85 @@ static void usage(const char *program_name)
+    }
+
+    /* Time up to first sync() call */
+    struct timespec postwrite;
```
Actually I would like the first stats to be calculated after the first sync() call, not before.

francescolavra commented on test/runtime/write.c:
```diff
@@ -617,6 +617,85 @@ static void usage(const char *program_name)
+        perror("fs_stress_test: clock_gettime");
+        goto out_fail;
+    }
+    print_op_stats("sync & delete", &postwrite, &postsync, 0);
```
Should be called just "delete", not "sync & delete"

francescolavra commented on test/runtime/write.c:
```diff
@@ -688,6 +767,8 @@ int main(int argc, char **argv)
     }
 
+    fs_stress_test();
+
```
This should be under the "basic" tests, so this function call should be in the `if (op == WRITE_OP_ALL || op == WRITE_OP_BASIC_ONLY)` block, a few lines above

uma-sethuraman commented:
@francescolavra we updated the test based on the changes requested. Is there anything else we should change?
